### PR TITLE
Chore: Migrate NotificationPreferencesView to hooks

### PR DESCRIPTION
**Pull request by GleidsonDaniel**
## Proposed changes

Migrate NotificationPreferencesView to hooks and improves layout and user feedback 
- It is no longer necessary to change screens to select some options, they are via actionSheet.
- Now when an error occurs when changing the option, it reverts to the old option.

## How to test or reproduce

- Try changing some option and see if it will actually replicate on the web or in tests.
- Force an error (too many requests for example) and see the error message and the go back option.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] Improvement (non-breaking change which improves a current function)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Documentation update (if none of the other choices apply)

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/RocketChat/Rocket.Chat/blob/develop/.github/CONTRIBUTING.md#contributing-to-rocketchat) doc
- [x] I have signed the [CLA](https://cla-assistant.io/RocketChat/Rocket.Chat.ReactNative)
- [x] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works (if applicable)
- [ ] I have added necessary documentation (if applicable)
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: app/containers/ActionSheet/Item.tsx [hunk 1]
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable no-nested-ternary */
```
Comment (diegolmello) [inline app/containers/ActionSheet/Item.tsx:1]: I don't think this is needed.

File: app/containers/ActionSheet/styles.ts [hunk 2]
```diff
@@ -63,5 +62,6 @@ export default StyleSheet.create({
 	},
 	rightContainer: {
 		paddingLeft: 12
-	}
+	},
+	none: { width: 20 }
```
Comment (diegolmello) [inline app/containers/ActionSheet/styles.ts:66]: You're not using it

File: app/views/NotificationPreferencesView/index.tsx [hunk 3]
```diff
@@ -1,282 +1,234 @@
...
-					this.state.room = changes;
-				}
-			});
-		}
-	}
-
-	componentDidMount() {
-		this.mounted = true;
-	}
+const RenderListPicker = React.memo(
```
Comment (diegolmello) [inline app/views/NotificationPreferencesView/index.tsx:29]: `room` is an object and `onChangeValue` is a function, so React.memo isn't really doing anything.

File: app/views/NotificationPreferencesView/index.tsx [hunk 4]
```diff
@@ -1,282 +1,234 @@
...
-			});
+			// @ts-ignore
+			logEvent(events[`NP_${key.toUpperCase()}`]);
+			await Services.saveNotificationSettings(rid, params);
 		} catch (e) {
 			// @ts-ignore
 			logEvent(events[`NP_${key.toUpperCase()}_F`]);
 			log(e);
+			onError();
+			showErrorAlertWithEMessage(e);
```
Comment (diegolmello) [inline app/views/NotificationPreferencesView/index.tsx:113]: This is confusing.
When is it calling `onError`?
Will it do both `onError` and `showErrorAlertWithEMessage`?

File: app/views/NotificationPreferencesView/index.tsx [hunk 5]
```diff
@@ -1,282 +1,234 @@
...
+const RenderSwitch = ({ preference, room, onChangeValue }: IBaseParams) => {
+	const [switchValue, setSwitchValue] = useState(!room[preference]);
+	return (
+		<Switch
+			value={switchValue}
+			testID={preference}
+			trackColor={SWITCH_TRACK_COLOR}
+			onValueChange={value => {
+				onChangeValue(preference, { [preference]: switchValue ? '1' : '0' }, () => setSwitchValue(switchValue));
+				setSwitchValue(value);
```
Comment (diegolmello) [inline app/views/NotificationPreferencesView/index.tsx:83]: Should we call `setSwitchValue` twice?

File: app/views/NotificationPreferencesView/index.tsx [hunk 6]
```diff
@@ -1,282 +1,234 @@
...
-interface INotificationPreferencesViewProps {
-	navigation: StackNavigationProp<ChatsStackParamList, 'NotificationPrefView'>;
-	route: RouteProp<ChatsStackParamList, 'NotificationPrefView'>;
-	theme: TSupportedThemes;
-	serverVersion: string | null;
-}
-
-interface INotificationPreferencesViewState {
+interface IBaseParams {
+	preference: string;
```
Comment (diegolmello) [inline app/views/NotificationPreferencesView/index.tsx:24]: I think we could actually type this.
This preference is an abstraction of both `OPTIONS` or `IRoomNotifications`.

For the first one, you could
- use `keyof typeof OPTIONS`
- remove [this nonsense](https://github.com/RocketChat/Rocket.Chat.ReactNative/blob/60f781d20e24ca92f9cd2c5dd9ef41de16fd0767/app/views/NotificationPreferencesView/options.ts#L7)
- remove the assertion `as IOptionsField` 
- Everything `RenderListPicker` is going to be inferred

Apply something like that to `IRoomNotifications` and we're good.

**Review by diegolmello — DISMISSED** [review]

File: app/views/NotificationPreferencesView/index.tsx [hunk 7]
```diff
@@ -1,282 +1,236 @@
...
+
+const RenderSwitch = ({ preference, room, onChangeValue }: IBaseParams) => {
+	const [switchValue, setSwitchValue] = useState(!room[preference]);
+	return (
+		<Switch
+			value={switchValue}
+			testID={preference as string}
+			trackColor={SWITCH_TRACK_COLOR}
+			onValueChange={value => {
+				onChangeValue(preference, { [preference]: switchValue ? '1' : '0' }, () => setSwitchValue(switchValue));
```
Comment (reinaldonetof) [inline app/views/NotificationPreferencesView/index.tsx:84]: WDYT to use named params?

File: app/views/NotificationPreferencesView/index.tsx [hunk 8]
```diff
@@ -1,282 +1,236 @@
...
+			title: I18n.t('Notification_Preferences')
+		});
+	}, []);
 
-	saveNotificationSettings = async (key: string, value: string | boolean, params: IRoomNotifications) => {
-		// @ts-ignore
-		logEvent(events[`NP_${key.toUpperCase()}`]);
-		const { room } = this.state;
-		const db = database.active;
+	const rid = route.params?.rid ?? '';
```
Comment (reinaldonetof) [inline app/views/NotificationPreferencesView/index.tsx:102]: Is this really needed? Is there any possibility the rid be undefined?

**Review by reinaldonetof — CHANGES_REQUESTED** [review]

File: app/views/NotificationPreferencesView/index.tsx (hunk 7 above)
Comment (GleidsonDaniel) [reply app/views/NotificationPreferencesView/index.tsx:84]: named params?

**Review by reinaldonetof — APPROVED** [review]

**Pull request merged**